src/lib/protocols/mdns/Publisher.cpp

```diff
@@ -0,0 +1,171 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "Publisher.h"
+
+#include <inttypes.h>
+
+#include "lib/protocols/mdns/platform/Mdns.h"
+#include "lib/support/logging/CHIPLogging.h"
+#include "platform/CHIPDeviceBuildConfig.h"
+#include "platform/CHIPDeviceLayer.h"
+#include "support/CodeUtils.h"
+#include "support/ErrorStr.h"
+#include "support/RandUtils.h"
+
+namespace chip {
+namespace Protocols {
+namespace Mdns {
+
+CHIP_ERROR Publisher::Init()
+{
+    CHIP_ERROR error;
+
+    mUnprovisionedInstanceName = GetRandU64();
+    SuccessOrExit(error = ChipMdnsInit(HandleMdnsInit, HandleMdnsError, this));
+    SuccessOrExit(error = SetupHostname());
+exit:
+    return error;
+}
+
+void Publisher::HandleMdnsInit(void * context, CHIP_ERROR initError)
+{
+    Publisher * publisher = static_cast<Publisher *>(context);
+
+    if (initError == CHIP_NO_ERROR)
+    {
+        publisher->mInitialized = true;
+    }
+    else
+    {
+        ChipLogError(Discovery, "mDNS initialization failed with %s", chip::ErrorStr(initError));
+    }
+}
+
+void Publisher::HandleMdnsError(void * context, CHIP_ERROR initError)
+{
+    ChipLogError(Discovery, "mDNS error: %s", chip::ErrorStr(initError));
+}
+
+CHIP_ERROR Publisher::StartPublishDevice(chip::Inet::InterfaceId interface)
+{
+    CHIP_ERROR error;
+
+    if (chip::DeviceLayer::ConfigurationMgr().IsFullyProvisioned() != mIsPublishingProvisionedDevice)
+    {
+        SuccessOrExit(error = StopPublishDevice());
+        // Set hostname again in case the mac address changes when shifting from soft-AP to station
+        SuccessOrExit(error = SetupHostname());
+    }
+    mIsPublishingProvisionedDevice = chip::DeviceLayer::ConfigurationMgr().IsFullyProvisioned();
+
+    if (mIsPublishingProvisionedDevice)
+    {
+        error = PublishProvisionedDevice(interface);
+    }
+    else
+    {
+        error = PublishUnprovisionedDevice(interface);
+    }
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::SetupHostname()
+{
+    uint8_t mac[6];
+    char hostname[13]; // Hostname will be the hex representation of mac.
+    CHIP_ERROR error;
+
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetPrimaryWiFiMACAddress(mac));
+    for (size_t i = 0; i < sizeof(mac); i++)
+    {
+        snprintf(&hostname[i * 2], sizeof(hostname) - i * 2, "%02X", mac[i]);
+    }
+    SuccessOrExit(error = ChipMdnsSetHostname(hostname));
+
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::PublishUnprovisionedDevice(chip::Inet::InterfaceId interface)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    MdnsService service;
+    uint16_t discriminator;
+    uint16_t vendorID;
+    uint16_t productID;
+    char discriminatorBuf[5];
+    char vendorProductBuf[10];
```

Comment:
Is there a constexpr StringLength in pw_string? I think so. It allows doing things like:

`char vendorProductBuf[StringLength("FFFF+FFFF") + 1];` to size strings in a way that it's obvious why they are sized that way.

Reply:
We are not using pigweed in the libCHIP code for now. I can add some comments on how these length are generated.